lib/datasource/index.ts

```diff
@@ -55,25 +59,139 @@ const datasources: Record<string, Datasource> = {
+    if (
+      getHostsFromRulesGithub === undefined ||
+      !getHostsFromRulesGithub.includes(URL.parse(url).hostname)
+    ) {
+      extraBaseUrls.push('github.com', 'gist.github.com');
+    } else {
+      getHostsFromRulesGithub.forEach(host => {
+        extraBaseUrls.push(host, `gist.${host}`);
+      });
+    }
+    url = parse(url, {
+      extraBaseUrls,
+    });
+    if (url !== null || url !== undefined) {
+      return url;
+    }
+  }
+  if (parsedUrl.protocol && parsedUrl.host && parsedUrl.path) {
+    let tmpProtocol: string;
+    if (parsedUrl.protocol !== 'http:') {
+      tmpProtocol = 'https:';
+    } else {
+      tmpProtocol = 'http:';
+    }
+    const getHostsFromRulesGitlab =
+      hostRules.hosts({ hostType: 'gitlab' }) || [];
+    const getHostsFromRulesBitbucket =
+      hostRules.hosts({ hostType: 'bitbucket' }) || [];
+    const getHostsFromRulesBitbucketServer =
+      hostRules.hosts({ hostType: 'bitbucket-server' }) || [];
+    const gitInfo = hostedGitInfo.fromUrl(url) || null;
+    let tmpurl: string;
+    if (gitInfo) {
+      tmpurl = gitInfo.browse({ noGitPlus: true, noCommittish: true });
+    }
+    if (
+      getHostsFromRulesGitlab.includes(parsedUrl.host) ||
+      parsedUrl.host === 'gitlab.com'
+    ) {
+      if (gitInfo && gitInfo.type === 'gitlab') {
+        url = tmpurl;
+      } else {
+        url = `${tmpProtocol}//${parsedUrl.host}${parsedUrl.path
+          .replace(RegExp('/api/v[3|4]/'), '')
+          .replace(':', '')
+          .replace(RegExp('.git$'), '')
+          .split('/')
+          .join('/')
+          .replace(RegExp('/$'), '')}`;
+      }
+    } else if (
+      getHostsFromRulesBitbucket.includes(parsedUrl.host) ||
+      parsedUrl.host === 'bitbucket.org' ||
+      getHostsFromRulesBitbucketServer.includes(parsedUrl.host)
+    ) {
+      if (gitInfo && gitInfo.type === 'bitbucket') {
+        url = tmpurl;
+      } else {
+        url = `${tmpProtocol}//${parsedUrl.host}${parsedUrl.path
+          .replace(':', '')
+          .split('/')
+          .slice(0, 3)
+          .join('/')
+          .replace(RegExp('.git$'), '')
+          .replace(RegExp('/$'), '')}`;
+      }
+    }
+    return url.replace(RegExp('.git$'), '').replace(RegExp('/$'), '');
+  }
+  /* todo add azure support. */
+  return null;
+}
+export async function fetchReleases(
   config: PkgReleaseConfig
 ): Promise<ReleaseResult | null> {
   const { datasource } = config;
+  // istanbul ignore if
   if (!datasource) {
     logger.warn('No datasource found');
-    return null;
   }
   if (!datasources[datasource]) {
     logger.warn('Unknown datasource: ' + datasource);
     return null;
   }
   const dep = await datasources[datasource].getPkgReleases(config);
   addMetaData(dep, datasource, config.lookupName);
+  if (dep && Object.entries(dep).length !== 0 && dep.sourceUrl !== undefined) {
+    const tmpSourceUrl: string | null = baseUrlLegacyMassager(dep.sourceUrl);
```

Comment:
```suggestion
    const tmpSourceUrl = baseUrlLegacyMassager(dep.sourceUrl);
```

type should come from `baseUrlLegacyMassager`return type.